### PR TITLE
rkt: Refactor version check with api-service. Add tests.

### PR DESCRIPTION
**Pull request by yifan-gu**

### CODE REVIEW

**Comment by yifan-gu** [comment]
Depends on #17527 

As api service is merged into rkt. It's time for refactoring the rkt code in kublet to use the clear defined interfaces.

/cc @dchen1107 @yujuhong @timstclair

**Comment by yujuhong** [comment]
The unit tests all failed because the builders don't have the required minimum version for systemd. You'll need to mock the getSystemdVersion function call, or use/extend the FakeOS interface.

```
<autogenerated>:31: 
                        
	Error Trace:	rkt_test.go:105
	
	Error:		Not equal: &errors.errorString{s:"rkt: systemd version is too old, requires at least 219"} (expected)
		        != <nil> (actual)
	
	Messages:	test case #0
	
<autogenerated>:31: 
                        
	Error Trace:	rkt_test.go:105
	
	Error:		Not equal: &errors.errorString{s:"rkt: systemd version is too old, requires at least 219"} (expected)
		        != <nil> (actual)
	
	Messages:	test case #1
	
<autogenerated>:31: 
                        
	Error Trace:	rkt_test.go:105
	
	Error:		Not equal: &errors.errorString{s:"rkt: systemd version is too old, requires at least 219"} (expected)
		        != &errors.errorString{s:"rkt: binary version is too old(1.2.3+git), requires at least 1.2.4"} (actual)
	
	Messages:	test case #2
	
<autogenerated>:31: 
                        
	Error Trace:	rkt_test.go:105
	
	Error:		Not equal: &errors.errorString{s:"rkt: systemd version is too old, requires at least 219"} (expected)
		        != &errors.errorString{s:"rkt: Appc version is too old(1.2.4+git), requires at least 1.2.5"} (actual)
	
	Messages:	test case #3
	
<autogenerated>:31: 
                        
	Error Trace:	rkt_test.go:105
	
	Error:		Not equal: &errors.errorString{s:"rkt: systemd version is too old, requires at least 219"} (expected)
		        != &errors.errorString{s:"rkt: API version is too old(1.2.6-alpha), requires at least 1.2.6"} (actual)
	
	Messages:	test case #4
	
<autogenerated>:31: 
                        
	Error Trace:	rkt_test.go:105
	
	Error:		Not equal: &errors.errorString{s:"rkt: systemd version is too old, requires at least 219"} (expected)
		        != &errors.errorString{s:"rkt: API version is too old(1.2.6-alpha), requires at least 1.2.7"} (actual)
```

**Comment by yifan-gu** [comment]
@yujuhong k.. thanks. Will update

**Comment by yifan-gu** [comment]
Was trying not to mock the systemd interface as I planned to pull that out into a package. But anyway let's do it for now.

**Comment by yifan-gu** [comment]
cc @jonboulle

File: pkg/kubelet/rkt/rkt.go [hunk 1]
```diff
@@ -53,10 +55,11 @@ import (
 const (
 	RktType = "rkt"
 
-	acVersion             = "0.7.1"
-	minimumRktVersion     = "0.9.0"
-	recommendRktVersion   = "0.9.0"
-	systemdMinimumVersion = "219"
+	minimumAppcVersion     = "0.7.1"
+	minimumRktBinVersion   = "0.9.0"
+	recommendRktBinVersion = "0.9.0"
```
Comment (jonboulle) [inline pkg/kubelet/rkt/rkt.go]: recommended

File: pkg/kubelet/rkt/rkt.go [hunk 2]
```diff
@@ -158,28 +168,13 @@ func New(config *Config,
 		rkt.imagePuller = kubecontainer.NewImagePuller(recorder, rkt, imageBackOff)
 	}
 
-	// Test the rkt version.
-	version, err := rkt.Version()
-	if err != nil {
+	if err := rkt.checkVersion(minimumRktBinVersion, recommendRktBinVersion, minimumAppcVersion, minimumRktApiVersion, minimumSystemdVersion); err != nil {
+		// TODO(yifan): Latest go-systemd version have the ability to close the
+		// dbus connection. However the 'docker/libcontainer' package is using
```
Comment (jonboulle) [inline pkg/kubelet/rkt/rkt.go]: can we bump libcontainer?

File: pkg/kubelet/rkt/rkt.go [hunk 3]
```diff
@@ -114,21 +127,16 @@ func New(config *Config,
 	imageBackOff *util.Backoff,
 	serializeImagePulls bool,
 ) (*Runtime, error) {
-	systemdVersion, err := getSystemdVersion()
-	if err != nil {
-		return nil, err
-	}
-	result, err := systemdVersion.Compare(systemdMinimumVersion)
+	// Create dbus connection.
+	systemd, err := newSystemd()
```
Comment (jonboulle) [inline pkg/kubelet/rkt/rkt.go]: did you forget to include a file in this diff?

File: pkg/kubelet/rkt/rkt.go (hunk 2 above)
Comment (jonboulle) [inline pkg/kubelet/rkt/rkt.go]: https://github.com/opencontainers/runc/pull/408

Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: @jonboulle We can submit a PR to libcontainer, but I don't what that block this.

File: pkg/kubelet/rkt/rkt.go (hunk 3 above)
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: I don't know what's happening here, but if you click the second commit, it shows up
https://github.com/yifan-gu/kubernetes/commit/da3a66eaa9fadb8d535b879360f0b5873a1f53dd

**Comment by yifan-gu** [comment]
@jonboulle Updated, wanna take another look?

**Comment by jonboulle** [comment]
LGTM, but we need to figure out the go-systemd/libcontainer dependency thing so we can get that bumped.

**Comment by yifan-gu** [comment]
@jonboulle I think I know how to fix that, just need to change the libcontainer to be the one inside runc repo.

**Comment by yifan-gu** [comment]
Turns out that cadvisor has a dependency on docker/libcontainer. Sent a PR to update that to use opencontainers/runc/libcontainer https://github.com/google/cadvisor/pull/973

**Comment by jonboulle** [comment]
@yifan-gu 
```rkt: Refactor verstion check with api-service. …	```
--> ver**s**ion

**Comment by yifan-gu** [comment]
@jonboulle Fixed title. For the dependency bump, we will wait https://github.com/google/cadvisor/pull/966 and https://github.com/google/cadvisor/pull/974 gets merged in cadvisor.

**Comment by yifan-gu** [comment]
@jonboulle I think we should do the version bump later? Does that sgty?

**Comment by jonboulle** [comment]
Yes, sgtm
:+1:

Yifan Gu <notifications@github.com> schrieb am Mo., 23. Nov. 2015 17:02:

**Comment by yifan-gu** [comment]
@dchen1107 @yujuhong @Random-Liu wanna add LGTM label to make this in quicker?

**Comment by yujuhong** [comment]
@k8s-bot test this

**Comment by yujuhong** [comment]
LGTM.

**Comment by yujuhong** [comment]
Applied the label, but no guarantee that it'd be quicker since we're all at merge bot's mercy.

**Comment by yifan-gu** [comment]
@yujuhong Thanks :)

**Pull request merged**